src/test/portmidicontroller_test.cpp

```diff
@@ -0,0 +1,480 @@
+        m_outputDeviceInfo.opened = 0;
+
+        m_pController.reset(new MockPortMidiController(&m_inputDeviceInfo,
+                                                       &m_outputDeviceInfo,
+                                                       0, 0));
+        m_pController->setPortMidiInputDevice(m_mockInput);
+        m_pController->setPortMidiOutputDevice(m_mockOutput);
+    }
+
+    void openDevice() {
+        m_pController->open();
+    }
+
+    void closeDevice() {
+        m_pController->close();
+    }
+
+    void pollDevice() {
+        m_pController->poll();
+    }
+
+    PmDeviceInfo m_inputDeviceInfo;
+    PmDeviceInfo m_outputDeviceInfo;
+    MockPortMidiDevice* m_mockInput;
+    MockPortMidiDevice* m_mockOutput;
+    QScopedPointer<MockPortMidiController> m_pController;
+};
+
+PmEvent MakeEvent(PmMessage message, PmTimestamp timestamp) {
+    PmEvent event;
+    event.message = message;
+    event.timestamp = timestamp;
+    return event;
+}
+
+MATCHER_P(ByteArrayEquals, value,
+          "Checks that the non-NULL terminated argument array exactly equals "
+          "the provided byte container.") {
+    for (int i = 0; i < value.size(); ++i) {
+        if (arg[i] != value.at(i))
+            return false;
+    }
+    return true;
+}
+
+TEST_F(PortMidiControllerTest, OpenClose) {
+    Sequence input;
+    ON_CALL(*m_mockInput, isOpen())
+            .WillByDefault(Return(false));
+    EXPECT_CALL(*m_mockInput, openInput(MIXXX_PORTMIDI_BUFFER_LEN))
+            .InSequence(input)
+            .WillOnce(Return(pmNoError));
+    EXPECT_CALL(*m_mockInput, isOpen())
+            .InSequence(input)
+            .WillRepeatedly(Return(true));
+    EXPECT_CALL(*m_mockInput, close())
+            .InSequence(input)
+            .WillOnce(Return(pmNoError));
+
+    Sequence output;
+    ON_CALL(*m_mockOutput, isOpen())
+            .WillByDefault(Return(false));
+    EXPECT_CALL(*m_mockOutput, openOutput())
+            .WillOnce(Return(pmNoError));
+    EXPECT_CALL(*m_mockOutput, isOpen())
+            .InSequence(output)
+            .WillRepeatedly(Return(true));
+    EXPECT_CALL(*m_mockOutput, close())
+            .InSequence(output)
+            .WillOnce(Return(pmNoError));
+
+    openDevice();
+    EXPECT_TRUE(m_pController->isOpen());
+    closeDevice();
+    EXPECT_FALSE(m_pController->isOpen());
+};
+
+TEST_F(PortMidiControllerTest, WriteShort) {
+    Sequence output;
+    EXPECT_CALL(*m_mockOutput, isOpen())
+            .WillRepeatedly(Return(true));
+    EXPECT_CALL(*m_mockOutput, writeShort(0x403C90))
+            .InSequence(output)
+            .WillOnce(Return(pmNoError));
+    EXPECT_CALL(*m_mockOutput, writeShort(0xFFFFFF))
+            .InSequence(output)
+            .WillOnce(Return(pmBadData));
+    EXPECT_CALL(*m_mockOutput, writeShort(0x403C80))
+            .InSequence(output)
+            .WillOnce(Return(pmNoError));
+
+    m_pController->sendShortMsg(0x90, 0x3C, 0x40);
```

Comment:
Have you worked with google mock before? Reading this might be helpful:
https://github.com/google/googletest/blob/master/googlemock/docs/ForDummies.md

The expectations setup the patterns of calls we expect to observe in the mock objects.